packages/belos/src/BelosGmresPolyOp.hpp

```diff
@@ -100,31 +267,425 @@ namespace Belos {
+      lapack.POTRF( 'U', dim_+1, lhstemp.values(), lhstemp.stride(), &infoInt);
+      
+      if(autoDeg == false)
+      { 
+        status = false;
+        if(infoInt != 0)
+        {
+          std::cout << "BelosGmresPolyOp.hpp: LAPACK POTRF was not successful!!" << std::endl;
+          std::cout << "Error code: " << infoInt << std::endl; 
+        } 
+      } 
+      else
+      {
+        if(infoInt != 0)
+        {//Had bad factor.  Reduce poly degree.
+          dim_--;
+        } 
+        else
+        {
+          status = false;
+        } 
+      } 
+      if(status == false)
+      {
+        lhs = lhstemp;
+      } 
+    } 
+    if(dim_ == 0)
+    {
+      pCoeff_.shape( 1, 1);
+      pCoeff_(0,0) = SCT::one(); 
+      std::cout << "Poly Degree is zero.  No preconditioner created." << std::endl;
+    } 
+    else
+    {
+      pCoeff_.shape( dim_+1, 1);
+      //Get correct submatrix of AV:
+      Teuchos::Range1D rangeSub( 1, dim_+1);
+      Teuchos::RCP< const MV > AVsub = MVT::CloneView( *V, rangeSub);
+      
+      //Compute rhs (AV)^T V0
+      MVT::MvTransMv( SCT::one(), *AVsub, *V0, pCoeff_);
+      lapack.POTRS( 'U', dim_+1, 1, lhs.values(), lhs.stride(), pCoeff_.values(), pCoeff_.stride(), &infoInt);
+      if(infoInt != 0) 
+      {
+        std::cout << "BelosGmresPolyOp.hpp: LAPACK POTRS was not successful!!" << std::endl;
+        std::cout << "Error code: " << infoInt << std::endl; 
+      } 
+    } 
+  }
+
+  template <class ScalarType, class MV, class OP>
+  void GmresPolyOp<ScalarType, MV, OP>::generateArnoldiPoly()
+  {
+    std::string polyLabel = label_ + ": GmresPolyOp creation";
+
+    // Create a copy of the linear problem that has a zero initial guess and random RHS.
+    std::vector<int> idx(1,0);
+    Teuchos::RCP<MV> newX  = MVT::Clone( *(problem_->getLHS()), 1 );
+    Teuchos::RCP<MV> newB  = MVT::Clone( *(problem_->getRHS()), 1 );
+    MVT::MvInit( *newX, SCT::zero() );
+    if (randomRHS_) {
+      MVT::MvRandom( *newB );
+    }
+    else {
+      MVT::Assign( *(MVT::CloneView(*(problem_->getRHS()), idx)), *newB );
+    }
+    Teuchos::RCP<LinearProblem<ScalarType,MV,OP> > newProblem =
+      Teuchos::rcp( new LinearProblem<ScalarType,MV,OP>( problem_->getOperator(), newX, newB ) );
+    newProblem->setInitResVec( newB );
+    newProblem->setLeftPrec( problem_->getLeftPrec() );
+    newProblem->setRightPrec( problem_->getRightPrec() );
+    newProblem->setLabel(polyLabel);
+    newProblem->setProblem();
+    newProblem->setLSIndex( idx );
+
+    // Create a parameter list for the GMRES iteration.
+    Teuchos::ParameterList polyList;
+ 
+    // Tell the block solver that the block size is one.
+    polyList.set("Num Blocks",maxDegree_);
+    polyList.set("Block Size",1);
+    polyList.set("Keep Hessenberg", true);
+
+    // Create orthogonalization manager if we need to.  
+    if (ortho_.is_null()) {
```

Comment:
Please use Belos::OrthoManagerFactory, to avoid repeating code and to include TSQR in the list of supported OrthoManager types.

Reply:
See above.

Reply:
I'm not suggesting changing the default here, just removing some duplicated code.